lxd/instance/drivers/driver_lxc.go

```diff
@@ -1301,6 +1301,56 @@ func (d *lxc) initLXC(config bool) error {
 	return nil
 }
 
+var idmappedStorageMap map[unix.Fsid]idmap.IdmapStorageType = map[unix.Fsid]idmap.IdmapStorageType{}
+var idmappedStorageMapLock sync.Mutex
+
+// IdmappedStorage determines if the container can use idmapped mounts or shiftfs
+func (d *lxc) IdmappedStorage(path string) idmap.IdmapStorageType {
```

Comment:
I think something like this would be cleaner:

```go
func (d *lxc) IdmappedStorage(path string) idmap.IdmapStorageType {
        // Fallback mode.
        mode := idmap.IdmapStorageNone
        if d.state.OS.Shiftfs {
                mode = idmap.IdmapStorageShiftfs
        }

        // Check if idmapped mounts are supported.
        if !d.state.OS.LXCFeatures["idmapped_mounts"] {
                return mode
        }

        // Get filesystem type.
        buf := &unix.Statfs_t{}
        err := unix.Statfs(path, buf)
        if err != nil {
                logger.Errorf(errors.Wrapf(err, "Failed to statfs"))
                return mode
        }

        // Only do one filesystem check at a time.
        idmappedStorageMapLock.Lock()
        defer idmappedStorageMapLock.Unlock()

        // Look for existing results.
        val, ok := idmappedStorageMap[buf.Fsid]
        if ok {
                return val
        }

        // Perform the check.
        val := idmap.CanIdmapMount(path)
        if val {
                mode = idmap.IdmapStorageIdmapped
        }
        idmappedStorageMap[buf.Fsid] = val

        return mode
}
```

This makes it a bit shorter and also has a nicer fallback on failure.

Reply:
Why do you want to keep the mutex over something as expensive as testing a detached idmap mount? I explicitly avoided that.

Reply:
My thought was that since it's expensive we also don't want multiple of those happening in parallel, so keeping the mutex will avoid that and as there's a very high chance that all containers will be on the same fsid, keeping the mutex means we won't ever do duplicate checks.

Reply:
The cost of contending that mutex especially with multiple containers will be quite high. So I'm not so sure about that. It should probably be fine but I don't like to hold mutexes over multiple syscalls especially in go.

Reply:
@brauner is that because syscalls are relatively more expensive in Go? And are the syscalls you referring to here the detached idmap mounts? Would allowing these expensive operations to occur multiple times, and then discarding some of the results (in the case where multiple operations were started for the same fs type) be more expensive than blocking multiple requests on a mutex?

Reply:
I've also never heard of sync.Mutex being particularly expensive in Go. It's used very extensively to protect all kinds of structs and maps not only in LXD but across the Go standard library.

To me it seems like the actual idmapped test will be expensive (mounting stuff and doing some other things in cgo), more expensive than waiting for a futex (for any other attempt happening in parallel) and once that initial test will have cleared, everything after it should just be locking the map for an instant as they pull the boolean out and return.

Reply:
> @brauner is that because syscalls are relatively more expensive in Go? And are the syscalls you referring to here the detached idmap mounts? Would allowing these expensive operations to occur multiple times, and then discarding some of the results (in the case where multiple operations were started for the same fs type) be more expensive than blocking multiple requests on a mutex?

My point is that I don't like holding a mutex from go on in a multi-threaded environments across multiple system call, i.e. entering the kernel, coming back to userspace, etc. That all relies on futexes and they have had serious bugs. The context switching part makes me uneasy and in general this is pretty much an anti-design pattern to hold a mutex for that long.

Reply:
> I've also never heard of sync.Mutex being particularly expensive in Go. It's used very extensively to protect all kinds of structs and maps not only in LXD but across the Go standard library.

Mutexes aren't necessarily expensive per se. They become very expensive when they are _contended_ is the point.

Reply:
In general we're often being very liberal with them...

Reply:
> I think something like this would be cleaner:
> 
> ```go
> func (d *lxc) IdmappedStorage(path string) idmap.IdmapStorageType {
>         // Fallback mode.
>         mode := idmap.IdmapStorageNone
>         if d.state.OS.Shiftfs {
>                 mode = idmap.IdmapStorageShiftfs
>         }
> 
>         // Check if idmapped mounts are supported.
>         if !d.state.OS.LXCFeatures["idmapped_mounts"] {
>                 return mode
>         }
> 
>         // Get filesystem type.
>         buf := &unix.Statfs_t{}
>         err := unix.Statfs(path, buf)
>         if err != nil {
>                 logger.Errorf(errors.Wrapf(err, "Failed to statfs"))
>                 return mode
>         }
> 
>         // Only do one filesystem check at a time.
>         idmappedStorageMapLock.Lock()
>         defer idmappedStorageMapLock.Unlock()
> 
>         // Look for existing results.
>         val, ok := idmappedStorageMap[buf.Fsid]
>         if ok {
>                 return val
>         }
> 
>         // Perform the check.
>         val := idmap.CanIdmapMount(path)
>         if val {
>                 mode = idmap.IdmapStorageIdmapped
>         }
>         idmappedStorageMap[buf.Fsid] = val

This is buggy  btw.

Reply:
@brauner based on our discussion here and on IRC, although not a fix, one way we could reduce contention and the amount of syscalls being made is to "pre-warm" the cache for the most common scenarios when LXD starts. What I mean by this is that if we moved the lock and the map into the `storage` package and then provided access to it via some helper functions (so that the `instance` package can still access it as it needs to now), but also at storage pool init (at LXD start), each storage pool could (optionally based on if its suitable) run the idmap check and populate the cache map.

It wouldn't catch all of the scenarios, but it would allow the common ones (i.e containers directly using storage pool that is backed by a single filesystem type).